State Saving Refactor and support code (NGWPC-4213)

PhilMiller opened the pull request:
Restructuring to consolidate the time-dependent state of a simulation into a coherent object, `NgenSimulation`. That can then be the kernel of state saved and restored as needed.

## Additions

- Introduce a class `NgenSimulation` to encapsulate all of the objects and logic that make up a running simulation. Contents are largely moved from ngen `main()`
- Several `.cpp` files that contain code moved from corresponding header files
  * `Layer`
  * `parallel_utils`
  * `Formulation`
  * `Catchment_Formulation`

## Changes

- Nexus outflow CSV files are written from the driver code, not from the simulation code, so that when and how they're written can be controlled more readily
- `HY_Features_MPI::nexuses()` now returns just those owned by the local process, because that's what the existing call sites all wanted, and were post-processing to get
- Make various bit of logic `const`-safe
  * `Network`
  * `HY_Features` in part
- `HY_Features`: narrow dependence on `Formulation_Manager` definition to simplify linkage


## Testing

1. Automated tests all pass, including new construction smoke-test for `NgenSimulation` class
2. Comparison of outputs on repo-included example configurations against `development`, results matched exactly

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

PhilMiller committed:
WIP

PhilMiller committed:
WIP2

PhilMiller committed:
Fix confdig header include

PhilMiller committed:
Fix compilation errors

PhilMiller committed:
Move code defining Formulation and Catchment_Formulation to standalone cpp files

PhilMiller committed:
Drop nexus_outfiles entirely

PhilMiller committed:
SurfaceLayer: Remove unused private member variable nexus_ids

PhilMiller committed:
Refactor ngen main() to encapsulate state during execution

PhilMiller committed:
Move outflow vectors into NgenSimulation class

PhilMiller committed:
Some logic simplification

PhilMiller committed:
Make progress variable into a persistent class member 'simulation_step_' to eventually support restart

PhilMiller committed:
Tweaks towards routing integration

PhilMiller committed:
BMI Module Formulation: update to match revised serialization protocol

PhilMiller committed:
Layer.hpp: Refactor update_models into new Layer.cpp

PhilMiller committed:
Layer.cpp: Reformat moved code

PhilMiller committed:
Delete commented out code containing an earlier proposed implementation of passing flows to routing in-memory

PhilMiller committed:
SurfaceLayer: Simplify logic of skipping output of remote-owned nexuses

PhilMiller committed:
SurfaceLayer: Clean up comments

PhilMiller committed:
SurfaceLayer: Check for and report non-dendritic network

PhilMiller committed:
Layer: Simplify namespace usage and out-dent

PhilMiller committed:
main(): Add hooks for running integrated coastal TWL

PhilMiller commented:
To people added as reviewers - note that this is still in draft form. It's posted as a PR to ensure that AC for NGWPC-8300 are satisfied.

PhilMiller committed:
Privatize some global variables only used in ngen main()

PhilMiller committed:
Privatize another global variable only used in ngen main()

PhilMiller committed:
Privatize stringstream only used in ngen main() for log message construction

PhilMiller committed:
Privatize global variable mpi_rank, leavning none crossing object boundaries or used non-locally

PhilMiller committed:
Rename formulations to formulation_constructors to clarify its purpose

PhilMiller committed:
Merge remote-tracking branch 'ngwpc/development' into PhilMiller-4213-state-saving

PhilMiller committed:
Remove defunct references to deleted Routing_Py_Adapter

PhilMiller commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,253 @@
+}
+
+
+void NgenSimulation::advance_one_output_step()
```
My intention is to move the body of the above loop down here, so that we can separate out any control logic for output and state saving.

PhilMiller committed:
Fix compilation errors in routing code

idtodd commented on include/core/NgenSimulation.hpp:
```diff
@@ -0,0 +1,56 @@
+    std::vector<std::shared_ptr<ngen::Layer>> layers_;
+
+    // Routing data structured for t-route
+    std::unordered_map<std::string, int> catchment_indexes_;
```
Catchment results still aren't being used for anything in memory, so it might be best to remove these. The catchment results may get saved as CSV outputs, but those are used by external programs, not used by NGEN subprocesses like T-route.

idtodd replied:
Alternatively, the if we want to prepare for checkpointing, the way the array is built needs to be changed. Currently, the CSVs are generated as NGEN runs. We could use the `catchment_outflows_` vector to store the CSV output values as the simulation runs, then write the data at the end of the simulation.

PhilMiller replied:
That was roughly my thinking, regarding checkpointing. Though, given those files end up with other variables too besides the 'main output variable' that expresses runoff, maybe we do away with this anyway.

PhilMiller replied:
For the basic hind-casting use case, we don't need to handle interrupted runs' mangled output, like we do for the resilience case. My intention is to leave this in place for now. We can perhaps think of it as the placeholder for the more complete set of catchment outputs to be captured this way.

PhilMiller committed:
Fix most compilation issues with merged/refactored code, besides missing HY_Features instance

idtodd commented on src/NGen.cpp:
```diff
@@ -139,6 +122,16 @@ void ngen::exec_info::runtime_summary(std::ostream& stream) noexcept {
+    std::string REALIZATION_CONFIG_PATH   = "";
+    bool is_subdivided_hydrofabric_wanted = false;
+    std::string PARTITION_PATH = "";
+    int mpi_num_procs;
```
Should we give this a default value (maybe 1 or -1)? Since this value is now being shared outside this file (e.g., NgenSimulation), it feels like something that could cause problems in the future if we don't control its initial value.

idtodd commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,254 @@
+
+}
+
+void NgenSimulation::run_routing()
```
This function as a whole is still referring to local versions of some values instead of members.

idtodd commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,254 @@
+            }
+        }
+
+        if (mpi_rank_ == 0) {
```
The NGen main function could move the collected indexes and flows since it wouldn't need to use those vectors after running t-route, but I think it'd be unwise to change NgenSimulation's members as part of this function. Maybe a local variable that is set to the member by default but will be changed to the merged values when running with MPI.

PhilMiller replied:
This is a good point about leaving the already-computed catchment and nexus outflow vectors as-is. I'll make an adjustment.

PhilMiller replied:
Done. If we later find a need to also retain the gathered data, those local variables can be made member variables as well.

idtodd commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,254 @@
+#include <Logger.hpp>
+
+#if NGEN_WITH_ROUTING
+#include "bmi/Bmi_Py_Adapter.hpp"
```
```suggestion
#include "bmi/Bmi_Py_Adapter.hpp"
#include "utilities/parallel_utils.h"
```
Parallel utils is needed for t-route.

PhilMiller commented:
Sorry about some of those errors - I missed pushing a commit last night. Still need to decide whether/how to bring in the `HY_Features_MPI` instance for the "is this nexus remote" check, or to get rid of it.

idtodd commented on src/core/SurfaceLayer.cpp:
```diff
@@ -14,47 +11,44 @@ void ngen::SurfaceLayer::update_models(boost::span<double> catchment_outflows,
         int nexus_index = nexus_indexes[id];
         nexus_downstream_flows[nexus_index] += contribution_at_t;
 #endif // NGEN_WITH_ROUTING
-        if(nexus_outfiles[id].is_open()) {
```
I think it may be premature to remove the nexus file outputs. Currently, they take an insignificant amount of the total runtime to write, so it's not acting a bottleneck at the moment. I'm also worried that there are lots of other parts of the project that are quietly relying on those files existing, so this change should be in coordination with all parts. For instance, cal manager would fail [here](https://github.com/NGWPC/nwm-cal-mgr/blob/37724ba7dbc9884844a9296fb2cf9de6c7cce4ac/python/calib/src/calib/calibration_set.py#L139) if no nexus files were generated.

PhilMiller replied:
I'm going to move the writing to an all-at-once step after the simulation has run, so that in a case where a job is interrupted, there will be no partial file output, and no need to figure out where to resume output.

PhilMiller replied:
Done in just-pushed changes.

idtodd commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,256 @@
+        std::vector<double> receive_buffer(number_of_timesteps, 0.0);
+        for (int i = 0; i < all_nexus_ids.size(); ++i) {
+            std::string nexus_id = all_nexus_ids[i];
+            if (nexus_indexes_.find(nexus_id) != nexus_indexes_.end() && !features.is_remote_sender_nexus(nexus_id)) {
```
`features` will need to be passed in as a parameter to this function. Or some other mechanism for filtering out remote nexuses.

PhilMiller replied:
I'll do the simple thing, passing that in, for now. I'd like to consider whether it might make more sense to rip out the `HY_Features_MPI` code entirely, in favor of this. There are scalability reasons that might not be a good idea, but that would need a bit of pencil-and-paper analysis to say for sure.

PhilMiller committed:
Pass through HY_Features_MPI instance to fix last compilation error

PhilMiller committed:
Extract implementations of parallel_utils.hpp functions into a separate file to avoid multiple-definition linker errors

PhilMiller committed:
Move catchment/nexus index data structures into the NgenSimulation constructor

PhilMiller committed:
Add sensible default value for mpi_num_procs

PhilMiller committed:
Network: improve const-correctness

PhilMiller committed:
HY_Feature(_MPI): improve const-correctness

PhilMiller committed:
Simplify nexus feature enumeration, and encapsulate MPI-specific behavior

PhilMiller committed:
Restore nexus outflow file writing, as an end-of-simulation bulk process

PhilMiller commented:
I should at least introduce a standalone test somewhere that instantiates `NgenSimulation` and sets up to exercise it, separate from its use in the `ngen` driver binary.

PhilMiller committed:
NgenSimulation: Avoid modifying nexus index and downstream flow structures once calculated

PhilMiller committed:
Refactor: extract single-macro-step updating into a separate function

idtodd commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,263 @@
+
+        if (simulation_step_ + 1 < num_times) {
+            sim_time_->advance_timestep();
+            catchment_formulation_manager_->Simulation_Time_Object->advance_timestep();
```
This logic should probably be moved to `advance_one_output_step`. If the goal is to let external items call the advance function, it would seem confusing to have to then tweak the simulation's simulation time object after each call.

PhilMiller replied:
Besides the rename mentioned above, I'm also making it `private`

idtodd commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,263 @@
+    int mpi_rank,
+    int mpi_num_procs
+                               )
+    : catchment_formulation_manager_(formulation_manager)
```
Is there a reason we need `NgenSimulation` to have a reference to the formulation manger? If the governing idea of creating the class is to make it portable and reusable, tying it to a `Formulation_Manager` this way seems like it has too much responsibility. Most of the need in the current code references just the `Simulation_Time_Object`, so the formulation manager could choose whether to share its simulation time with the `NgenSimulation` or create a copy. The only other places it's needed by `NgenSimulation` is in the routing, but the check for whether to run is already done in `NGen.cpp`, and the T-Route config file could just be another parameter.

PhilMiller replied:
I was mostly a matter of expedience, really. The `Formulation_Manager` class is in many ways an incipient version of this more explicit `NgenSimulation` class.

I was somewhat trying to keep from blowing up my footprint with this PR, but maybe I should just go ahead with it. I'll reduce the dependence, probably to zero.

PhilMiller committed:
NgenSimulation: Privatize members, and add the necessary additional method

idtodd commented on src/core/NgenSimulation.cpp:
```diff
@@ -0,0 +1,263 @@
+    for (; simulation_step_ < num_times; simulation_step_++) {
+        // Make room for this output step's results
+        catchment_outflows_.resize(catchment_outflows_.size() + catchment_indexes_.size());
+        nexus_downstream_flows_.resize(nexus_downstream_flows_.size() + nexus_indexes_.size());
```
Is the shift to resizing each step to reduce the size of the saved state? A couple considerations given this change:
1) This logic also seems better to include in `advance_one_output_step` instead of here, .e.g, something could call the advance function but the size of the output vectors is not large enough to contain the results.
2) We may want to explicitly set the last `nexus_indxexes_.size()` elements to 0. Since the layer will add the nexus value instead of set it, backtracking in time and decreasing size could yield double counts of that time step.
3) `run_routing` should check the size of `nexus_downstream_flows_`. It was built assuming the size of the data would be correct, but I'm not sure it's safe to do that anymore.

PhilMiller replied:
Re 1: Would it seem more reasonable to you if I renamed that function `advance_models_one_output_step`, with responsibility for managing data structures and metadata kept here?

PhilMiller replied:
Re 2: Good catch, I'll add the explicit initialization argument

PhilMiller replied:
Re 3, I'll check that

PhilMiller replied:
I think these concerns are all addressed in my latest push.

PhilMiller committed:
Refactor: shift responsibilities out of Formulation_Manager

- Shift Simulation_Time construction up to main(), and ownership of
  the dynamic instance to NgenSimulation
- Break the rest of NgenSimulation's dependence on
  Formulation_Manager, by shifting the minimal interaction up to
  main()

PhilMiller committed:
Rename to NgenSimulation::advance_models_one_output_step to clarify responsibility

PhilMiller committed:
NgenSimulation: Document more substantial public member functions

PhilMiller committed:
Explicitly initialize flow arrays to 0.0, so we can be confident about accumulating into them

PhilMiller committed:
Check nexus_downstream_flows_ against expected size before passing it to t-route

PhilMiller committed:
Forward declare HY_Features(_MPI) in another commonly-included header

PhilMiller committed:
Correct a widespread typo seperator -> separator

PhilMiller commented:
Thanks for the detailed review. We'll need to hold off on merging at least temporarily while blocker bugs get addressed and the release branch is formed.

PhilMiller committed:
Merge remote-tracking branch 'ngwpc/development' into PhilMiller-4213-state-saving

PhilMiller committed:
Formulation_Manager_Test: Fix compilation errors from changed interfaces

PhilMiller committed:
HY_Features: Drop shared_ptr<Formulation_Manager> as a member, and forward declare it, to mitigate linker errors

PhilMiller committed:
HY_Features_MPI: Same Formulation_Manager dependence reduction

PhilMiller committed:
Layer: Fix knock-on effects of lifting HY_Features -> Formulation_Manager dependence

PhilMiller committed:
MultiLayerParserTest: Fix knock-on effects of Formulation_Manager interface changes

PhilMiller committed:
Reflect NGen::core's dependence on parallel utils and ultimately on MPI

PhilMiller committed:
NgenSimulation construction smoke test

PhilMiller committed:
Log coastal time like everything else

PhilMiller commented:
Need to hold off merging, noticed a failure in `Layer::simulation_time` member, maybe related.

PhilMiller committed:
Fix accidental aliasing mistake

PhilMiller commented:
Awaiting some output verification against `development` before hitting 'Merge'

PhilMiller commented:
Ok, I did some minimal tests comparing results against `development`. Merging.
